tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/common/BatchConfiguration.java

```diff
@@ -69,63 +71,48 @@ public static BatchConfiguration computeChunkSizes(TornadoExecutionContext conte
 
-                classObjects.add(componentType);
+                elementSizes.add(dataTypeSize.getSize());
                 inputSizes.add(totalSize);
-                if (classObjects.size() > 1) {
-                    throw new TornadoRuntimeException("[UNSUPPORTED] Input objects with different data types not currently supported");
-                }
-                if (inputSizes.size() > 1) {
-                    throw new TornadoRuntimeException("[UNSUPPORTED] Input objects with different sizes not currently supported");
-                }
-            } else if (o instanceof IntArray) {
-                totalSize = ((IntArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(int.class);
-            } else if (o instanceof FloatArray) {
-                totalSize = ((FloatArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(float.class);
-            } else if (o instanceof DoubleArray) {
-                totalSize = ((DoubleArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(double.class);
-            } else if (o instanceof LongArray) {
-                totalSize = ((LongArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(long.class);
-            } else if (o instanceof ShortArray) {
-                totalSize = ((ShortArray) o).getNumBytesWithoutHeader();
+            } else if (o instanceof TornadoNativeArray tornadoNativeArray) {
+                totalSize = tornadoNativeArray.getNumBytesWithoutHeader();
                 inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(short.class);
-            } else if (o instanceof ByteArray) {
-                totalSize = ((ByteArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(byte.class);
-            } else if (o instanceof CharArray) {
-                totalSize = ((CharArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(char.class);
+                byte elementSize = switch (tornadoNativeArray) {
+                    case IntArray _ -> DataTypeSize.INT.getSize();
+                    case FloatArray _ -> DataTypeSize.FLOAT.getSize();
+                    case DoubleArray _ -> DataTypeSize.DOUBLE.getSize();
+                    case LongArray _ -> DataTypeSize.LONG.getSize();
+                    case ShortArray _ -> DataTypeSize.SHORT.getSize();
+                    case ByteArray _ -> DataTypeSize.BYTE.getSize();
+                    case CharArray _ -> DataTypeSize.CHAR.getSize();
+                    default -> throw new TornadoRuntimeException("Unsupported array type: " + o.getClass());
```

Comment:
This is needed as TornadoNativeArray isn't sealed. Not sure if you considered it before, but I don't know if there is any reason for custom implementations of either TornadoNativeArray or any of the special array types.

Reply:
Sorry, I was referring to the `default` case. The type-specific implementations make sense, but they are extendable (I could write `MyCustomIntArray extends IntArray`) which most likely doesn't make sense, and TornadoNativeArray is extendable as well (I could write `MyCustomWhateverArray extends TornadoNativeArray`) which most likely doesn't make sense either.

My suggestion was to make TornadoNativeArray `sealed`, only allow the type-specific implementations as subtypes, and make the subtypes final. This way, the `default` case here could be removed, and introducing a new type-specific implementation (e.g. `HalfFloatArray`) would be a java compiler error in this switch rather than only failing at runtime as soon as someone wants to use it. But this is something that can be addressed separately from this PR I guess.

Reply:
This is great feedback. I think makes sense. Let me analyze it first with the current plans and use cases for TornadoVM.